### PR TITLE
ADE SP Ubuntu 20.04 Support

### PR DESCRIPTION
**Pull request by ejarvi**
- adds support for Ubuntu 20.04

### CODE REVIEW

File: VMEncryption/main/oscrypto/ubuntu_2004/Ubuntu2004EncryptionStateMachine.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+scriptdir = os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))
+maindir = os.path.abspath(os.path.join(scriptdir, '../../'))
+sys.path.append(maindir)
+sixdir = os.path.abspath(os.path.join(scriptdir, '../../six'))
+sys.path.append(sixdir)
+transitionsdir = os.path.abspath(os.path.join(scriptdir, '../../transitions'))
+sys.path.append(transitionsdir)
+
+from oscrypto import *
```
Comment (vermashi) [inline VMEncryption/main/oscrypto/ubuntu_2004/Ubuntu2004EncryptionStateMachine.py:32]: Can we avoid * imports here?

Comment (ejarvi) [reply VMEncryption/main/oscrypto/ubuntu_2004/Ubuntu2004EncryptionStateMachine.py:32]: Good catch

File: VMEncryption/main/oscrypto/ubuntu_2004/encryptstates/PrereqState.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            raise Exception("Ubuntu2004EncryptionStateMachine called for distro {0} {1}".format(distro_info[0],
+                                                                                                distro_info[1]))
+
+        self.context.distro_patcher.install_extras()
+
+        self._patch_walinuxagent()
+        self.command_executor.Execute('systemctl daemon-reload', True)
+
+        self._copy_key_script()        
+        self._snap_stop()
```
Comment (vermashi) [inline VMEncryption/main/oscrypto/ubuntu_2004/encryptstates/PrereqState.py:62]: gate this behind an 'if' statement if we decide to merge the state machines.

File: VMEncryption/main/oscrypto/ubuntu_2004/Ubuntu2004EncryptionStateMachine.py (hunk 1 above)
Comment (ejarvi) [reply VMEncryption/main/oscrypto/ubuntu_2004/Ubuntu2004EncryptionStateMachine.py:32]: Would like to fix these in all state machines, not just Ubuntu 20.04.  Created an issue to track separately https://github.com/Azure/azure-linux-extensions/issues/1322 and clean up all at once in a separate change.

**Comment by ejarvi** [comment]
@vermashi @vimish @aravindthoram please approve

**Review by vermashi — APPROVED** [review]

File: VMEncryption/main/oscrypto/ubuntu_2004/encryptscripts/crypt-ade-hook [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    prereqs)
+    prereqs
+    exit 0
+    ;;
+esac
+
+# This ADE hook script adds two mount points and /etc/fstab entries to initramfs
+# so the LUKS header and key file are available for the boot script to use prior 
+# to calling cryptsetup to open the underlying root device /dev/mapper/osencrypt
+# The boot script is: /usr/share/initramfs-tools/scripts/init-premount/crypt-ade
```
[the remark this answers is not in the document]
Comment (ejarvi) [reply VMEncryption/main/oscrypto/ubuntu_2004/encryptscripts/crypt-ade-hook]: Yes, the comment is referencing the corresponding boot script which it supports.

File: VMEncryption/main/oscrypto/ubuntu_2004/encryptscripts/crypt-ade-boot [hunk 4]
```diff
@@ -0,0 +1,19 @@
+#!/bin/sh
+set -x
+
+# init-premount script for ADE
+# relies on the initramfs environment to have been prepared such that 
+# /etc/fstab is ready to mount the boot partition and key volume that
+# cryptsetup uses to open encrypted root device /dev/mapper/osencrypt
+# The hook script is here: /usr/share/initramfs-tools/hooks/crypt-ade
```
[the remark this answers is not in the document]
Comment (ejarvi) [reply VMEncryption/main/oscrypto/ubuntu_2004/encryptscripts/crypt-ade-boot]: Yes, this note is referencing where the prerequisite hook script for this boot script is located.

File: VMEncryption/main/oscrypto/ubuntu_2004/encryptscripts/inject_luks_header.sh [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    prereqs)
+    prereqs
+    exit 0
+    ;;
+esac
+
+. /usr/share/initramfs-tools/hook-functions
+
+mkdir -p ${DESTDIR}/boot/luks
+copy_exec /boot/luks/osluksheader /boot/luks
```
[the remark this answers is not in the document]
Comment (ejarvi) [reply VMEncryption/main/oscrypto/ubuntu_2004/encryptscripts/inject_luks_header.sh]: outdated, this script was removed as result of earlier feedback, update encryption settings tests are passing on current changes

File: VMEncryption/main/oscrypto/ubuntu_2004/encryptscripts/crypt-ade-hook (hunk 3 above)
[the remark this answers is not in the document]
Comment (ejarvi) [reply VMEncryption/main/oscrypto/ubuntu_2004/encryptscripts/crypt-ade-hook]: Yes, the comment is referencing the corresponding boot script that it supports for informational purposes.

File: VMEncryption/main/oscrypto/ubuntu_2004/encryptscripts/crypt-ade-boot (hunk 4 above)
[the remark this answers is not in the document]
Comment (ejarvi) [reply VMEncryption/main/oscrypto/ubuntu_2004/encryptscripts/crypt-ade-boot]: outdated, comments are for informational purposes to reference the corresponding hook script

File: VMEncryption/main/oscrypto/ubuntu_2004/encryptscripts/crypt-ade-hook [hunk 6]
```diff
@@ -0,0 +1,30 @@
+#!/bin/sh
+# -*- mode: shell-script; indent-tabs-mode: nil; sh-basic-offset: 4; -*-
+# ex: ts=8 sw=4 sts=4 et filetype=sh
+#set -x
+
+PREREQS=""
```
[the remark this answers is not in the document]
Comment (ejarvi) [reply VMEncryption/main/oscrypto/ubuntu_2004/encryptscripts/crypt-ade-hook:6]: merged and consolidated in latest version

**Pull request merged**